libyul/optimiser/SyntacticalEquality.cpp

```diff
@@ -64,9 +64,8 @@ bool SyntacticallyEqual::expressionEqual(Identifier const& _lhs, Identifier cons
 }
 bool SyntacticallyEqual::expressionEqual(Literal const& _lhs, Literal const& _rhs)
 {
-	yulAssert(validLiteral(_lhs), "Invalid lhs literal during syntactical equality check");
-	yulAssert(validLiteral(_rhs), "Invalid rhs literal during syntactical equality check");
-
+	assert(validLiteral(_lhs));
+	assert(validLiteral(_rhs));
```

Comment:
Is this really significant? `validLiteral` still is being called twice.

Reply:
I'd myself just have removed these entirely. Any assertion that we deem ok to just do non-release we might as well not do at all in our context. But whatever, no need to keep this waiting.

Reply:
> Any assertion that we deem ok to just do non-release we might as well not do at all in our context.

I don't agree. The built-in `assert` is perfectly fine as a baseline for sanity checks and even when disabled still adds value by documenting the assumptions. Having asserts checked in release builds is definitely nice to have and saved us many times, but when they're too costly I'd still rather fall back to using `assert` than remove them completely.

> The method `validLiteral` is nontrivial and should not be called (twice!) on every call to the equality comparison.
> It might be desirable to re-introduce the check in a more suitable place.

We're asserting validity here because `==` does not compare the whole `LiteralValue`. This is fine for valid literals but not in general and I don't think that's obvious.

A better place for these checks would be the implementation of `==`, but that would introduce a circular dependency. And would not solve the performance issue.